source/game_sa/CWeapon.cpp

```diff
@@ -283,9 +362,45 @@ float CWeapon::EvaluateTargetForHeatSeekingMissile(CEntity* entity, CVector& pos
     return plugin::CallAndReturn<float, 0x73E560, CEntity*, CVector&, CVector&, float, bool, CEntity*>(entity, posn, direction, distanceMultiplier, fromVehicle, lastEntity);
 }
 
-// Converted from thiscall void CWeapon::DoWeaponEffect(CVector origin,CVector target) 0x73E690
+// 0x73E690
 void CWeapon::DoWeaponEffect(CVector origin, CVector target) {
-    plugin::CallMethod<0x73E690, CWeapon*, CVector, CVector>(this, origin, target);
+    char fxName[32];
+
+    switch (m_nType) {
+    case WEAPON_FLAMETHROWER:
+        strcpy(fxName, "flamethrower");
+        break;
+    case WEAPON_EXTINGUISHER:
+        strcpy(fxName, "extinguisher");
+        break;
+    case WEAPON_SPRAYCAN:
+        strcpy(fxName, "spraycan");
+        break;
+    default:
+        return StopWeaponEffect();
+    }
+
+    RwMatrix* mat = RwMatrixCreate();
+    g_fx.CreateMatFromVec(mat, &origin, &target);
+
+    if (m_pFxSystem) {
+        m_pFxSystem->SetMatrix(mat);
+        m_pFxSystem->SetConstTime(1, 1.0f);
+    } else {
+        CVector origin(0.0f, 0.0f, 0.0f);
+
+        m_pFxSystem = g_fxMan.CreateFxSystem(fxName, &origin, mat, false);
+
+        if (!m_pFxSystem)
+            return;
```

Comment:
`RwMatrix` not destroyed
```suggestion
        if (!m_pFxSystem) {
            RwMatrixDestroy(mat);
            return;
        }
```